modules/markdown/markdown.go

```diff
@@ -261,48 +266,259 @@ func RenderIssueIndexPattern(rawBytes []byte, urlPrefix string, metas map[string
 	return rawBytes
 }
 
-// RenderSha1CurrentPattern renders SHA1 strings to corresponding links that assumes in the same repository.
-func RenderSha1CurrentPattern(rawBytes []byte, urlPrefix string) []byte {
-	return []byte(Sha1CurrentPattern.ReplaceAllStringFunc(string(rawBytes[:]), func(m string) string {
-		if com.StrTo(m).MustInt() > 0 {
-			return m
+// renderSha1CurrentPattern renders SHA1 strings to corresponding links that assumes in the same repository.
+func renderSha1CurrentPattern(rawBytes []byte, repoLink, urlPrefix string) []byte {
+	ms := Sha1CurrentPattern.FindAllSubmatch(rawBytes, -1)
+	for _, m := range ms {
+		all := m[0]
+		if com.StrTo(all).MustInt() > 0 {
+			continue
+		}
+		rawBytes = bytes.Replace(rawBytes, all, []byte(fmt.Sprintf(
+			`<a href="%s/commit/%s">%s</a>`, repoLink, all, base.ShortSha(string(all)))), -1)
+	}
+	return rawBytes
+}
+
+// renderFullSha1Pattern renders SHA containing URLs
+func renderFullSha1Pattern(rawBytes []byte, repoLink, urlPrefix string) []byte {
+	ms := AnySHA1Pattern.FindAllSubmatch(rawBytes, -1)
+	for _, m := range ms {
+		all := m[0]
+		// author := m[1]
+		// repoName := m[2]
+		ltype := "src"
+		itemType := m[3]
+		if string(itemType) == "commit" {
+			ltype = "commit"
+		}
+		sha := m[4]
+		var path []byte
+		if len(m) > 5 && len(m[5]) > 0 {
+			path = m[5]
+		}
+		var line []byte
+		if len(m) > 6 && len(m[6]) > 0 {
+			line = m[6]
+		}
+		urlSuffix := ""
+		text := base.ShortSha(string(sha))
+		if path != nil {
+			value := string(path)
+			urlSuffix = "/" + value
+			text = value
+		}
+		if line != nil {
+			value := string(line)
+			urlSuffix += "#"
+			urlSuffix += value
+			text += " ("
+			text += value
+			text += ")"
+		}
+		rawBytes = bytes.Replace(rawBytes, all, []byte(fmt.Sprintf(
+			`<a href="%s/%s/%s%s">%s</a>`, repoLink, ltype, sha, urlSuffix, text)), -1)
+	}
+	return rawBytes
+}
+
+// renderFullIssuePattern renders issues-like URLs
+func renderFullIssuePattern(rawBytes []byte, repoLink, urlPrefix string) []byte {
```

Comment:
I found problems with issue pattern rendering.

Link `https://github.com/go-gitea/gitea/pull/186` renders to:
- Title: `#186`
- Preview URL: `http://localhost:3000/issues/186`
- After saved URL: `http://localhost:3000/andreynering/mylocalrepo/issues/186`

Link `https://github.com/go-gitea/gitea/pull/186/files` renders to:
- Title: `#186/files`
- Preview URL: `http://localhost:3000/issues/186`
- After saved URL: `http://localhost:3000/andreynering/mylocalrepo/issues/186`

The original URL should be preserved. It can refer to GitHub or another Gitea instance, as example.